treeherder/model/management/commands/cycle_data.py

```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
+                    signature_id=signature.id,
+                ).exists():
+                    signature.delete()
+        except MaxRuntimeExceeded as ex:
+            logger.warning(ex)
+
+    def _maybe_quit(self):
+        elapsed_runtime = datetime.now() - self.started_at
+
+        if self.max_runtime < elapsed_runtime:
+            raise MaxRuntimeExceeded('Max runtime for performance data cycling exceeded')
+
+    def _delete_in_chunks(self, strategy: RemovalStrategy):
+        any_successful_attempt = False
+
+        with connection.cursor() as cursor:
+            while True:
+                self._maybe_quit()
+
+                try:
+                    strategy.remove(using=cursor)
+                except Exception as ex:
+                    self.__handle_chunk_removal_exception(ex, cursor, any_successful_attempt)
+                    break
+                else:
+                    deleted_rows = cursor.rowcount
+
+                    if deleted_rows == 0 or deleted_rows == -1:
+                        break  # either finished removing all expired data or failed
+                    else:
+                        any_successful_attempt = True
+                        logger.warning(
+                            'Successfully deleted {} performance datum rows'.format(deleted_rows)
+                        )
+
+    def __handle_chunk_removal_exception(
+        self, exception, cursor: CursorWrapper, any_successful_attempt: bool
+    ):
+        msg = 'Failed to delete performance data chunk'
+        if hasattr(cursor, '_last_executed'):
+            msg = f'{msg}, while running "{cursor._last_executed}" query'
+        logger.warning(msg)
+
+        if any_successful_attempt is False:
+            raise NoDataCyclingAtAll from exception
+
+
+class RemovalStrategy(ABC):
+    @abstractmethod
+    def remove(self, using: CursorWrapper):
+        pass
+
+    @staticmethod
+    def fabricate_all_strategies(*args, **kwargs):
+        return [
+            MainRemovalStrategy(*args, **kwargs),
+            TryDataRemoval(*args, **kwargs),
+            # append here any new strategies
+            # ...
+        ]
+
+
+class MainRemovalStrategy(RemovalStrategy):
+    """
+    Removes `performance_datum` rows
+    that are at least 1 year old.
+    """
+
+    # WARNING!! Don't override this without proper approval!
+    CYCLE_INTERVAL = 365  # in days                        #
+    ########################################################
+
+    def __init__(self, chunk_size: int):
+        self._cycle_interval = timedelta(days=self.CYCLE_INTERVAL)
+        self._chunk_size = chunk_size
+        self._max_timestamp = datetime.now() - self._cycle_interval
+        self._manager = PerformanceDatum.objects
+
+    def remove(self, using: CursorWrapper):
+        chunk_size = self._find_ideal_chunk_size()
+        using.execute(
+            '''
+            DELETE FROM `performance_datum`
+            WHERE push_timestamp < %s
```

Comment:
Giving the name `_max_timestamp` I would use here like:
`WHERE push_timestamp <= %s`